pkg/cmd/experimental/diagnostics/cluster.go

```diff
@@ -0,0 +1,115 @@
+package diagnostics
+
+import (
+	"fmt"
+	"strings"
+
+	kclient "k8s.io/kubernetes/pkg/client"
+	clientcmd "k8s.io/kubernetes/pkg/client/clientcmd"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/clientcmd/api"
+	"k8s.io/kubernetes/pkg/util"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/client"
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+
+	clustdiags "github.com/openshift/origin/pkg/diagnostics/cluster"
+	"github.com/openshift/origin/pkg/diagnostics/types"
+)
+
+var (
+	AvailableClusterDiagnostics = util.NewStringSet("NodeDefinitions", "ClusterRegistry", "ClusterRouter")
+)
+
+func (o DiagnosticsOptions) buildClusterDiagnostics(rawConfig *clientcmdapi.Config) ([]types.Diagnostic, bool /* ok */, error) {
+	requestedDiagnostics := intersection(util.NewStringSet(o.RequestedDiagnostics...), AvailableClusterDiagnostics).List()
+	if len(requestedDiagnostics) == 0 { // no diagnostics to run here
+		return nil, true, nil // don't waste time on discovery
+	}
+
+	var clusterClient *client.Client
+	var kclusterClient *kclient.Client
+
+	clusterClient, kclusterClient, found, err := o.findClusterClients(rawConfig)
+	if !found {
+		o.Logger.Notice("noClustCtx", "No cluster-admin client config found; skipping cluster diagnostics.")
+		return nil, false, err
+	}
+
+	diagnostics := []types.Diagnostic{}
+	for _, diagnosticName := range requestedDiagnostics {
+		switch diagnosticName {
+		case "NodeDefinitions":
+			diagnostics = append(diagnostics, &clustdiags.NodeDefinitions{kclusterClient, clusterClient})
+		case "ClusterRegistry":
+			diagnostics = append(diagnostics, &clustdiags.ClusterRegistry{kclusterClient, clusterClient})
+		case "ClusterRouter":
+			diagnostics = append(diagnostics, &clustdiags.ClusterRouter{kclusterClient, clusterClient})
+
+		default:
+			return nil, false, fmt.Errorf("unknown diagnostic: %v", diagnosticName)
+		}
+	}
+	return diagnostics, true, nil
+}
+
+func (o DiagnosticsOptions) findClusterClients(rawConfig *clientcmdapi.Config) (*client.Client, *kclient.Client, bool, error) {
+	if o.ClientClusterContext != "" { // user has specified cluster context to use
+		if context, exists := rawConfig.Contexts[o.ClientClusterContext]; exists {
+			configErr := fmt.Errorf("Specified '%s' as cluster-admin context, but it was not found in your client configuration.", o.ClientClusterContext)
+			o.Logger.Error("discClustCtx", configErr.Error())
```

Comment:
If so - yeah, that's the key, not an error message. The idea is that automation could filter for particular error IDs and not have to read the text to decide whether it's important.